### ISSUE TITLE
configure/CMake + SIMD/pthreads

### ISSUE THREAD

**Issue by kripken** [emscripten-core/emscripten#11877]
Talking with @surma it sounds like configure and CMake may not work with new features like SIMD and pthreads. That is, if the developer wants to use SIMD, and the configure check runs on a node version which is too old to have it, the check would fail.

The question is, do configure scripts do such checks, and whether we need to do something for it. If we do, options might include making it easier to use newer node or v8, and/or giving clear errors for how to fix this or work around it, and docs.

If relevant, we may want to add configure tests to our test suite for this. I'm pretty sure the existing configure tests do not use SIMD or pthreads.

I'm not sure exactly which flags users would need for SIMD, and which ones a configure script would check. Presumably only the SSE emulation stuff would be relevant here? For pthreads, the normal `-pthread` flag would be relevant I assume.

cc @tlively

**Comment by surma** [comment]
As an anecdote: [I experimented with MozJPEG and SIMD autovectorization](https://github.com/GoogleChromeLabs/squoosh/commit/0e90f805a4df8c2186aac2c1c66b03f015705281). 

`./configure --enable-simd [... other config flags ...]` would fail as the Node used to run the test binaries would detect an unknown opcode (i.e. the SIMD instructions). To work around this, I manually installed a recent version (v14.7) of Node and wrote a custom `.emscripten` so I can pass `--experimental-wasm-simd` to it.

**Comment by tlively** [comment]
Sorry this slipped through the cracks in my inbox.

Writing a custom `.emscripten` that includes necessary engine flags is how I would work around this as well. It would be good to mention this in the SIMD documentation.

**Comment by surma** [comment]
> Writing a custom .emscripten that includes necessary engine flags is how I would work around this as well. It would be good to mention this in the SIMD documentation.

It’s a decent workaround that should be documented, but if we consider SIMD support stable, I’d aim for something with less friction for developers.

**Comment by danchitnis** [comment]
Hi, I was wondering if the Emscripten team is planning to implement some level of OpenMP functionality especially auto-vectorization in emcc? Since both SIMD and pthreads are available in WebAssembly, in theory this should be possible. This would be a significant boost to computational web apps especially since the Chrome team doesn't seem to have decided to implement auto-vectorization in V8 [mentioned here](https://stackoverflow.com/questions/63864497/is-there-anyway-to-get-node-js-and-v8-to-automatically-vectorize-simple-loops).

**Comment by tlively** [comment]
@danchitnis LLVM's normal autovectorization is enabled when SIMD is used (although it hasn't been tuned very much yet). I don't know of any concrete plans to look into OpenMP support specifically, and I also don't know how much work that would be.

**Comment by danchitnis** [comment]
@tlively yes I did eventually realize that the `-msimd128` flag existed. However, auto-vectorization without a report is not very helpful, since the developer needs to know the compiler's success in the vectorization effort. Regarding tools like OpenMP, there have been various competing technologies like OpenCL, SYCL and OpenMP, all trying to achieve similar goals, and rapidly changing their specifications.

I would say an effective approach is for emscripten to have a C++ wrapper commands and data types around [WASM SIMD](https://github.com/WebAssembly/simd/blob/master/proposals/simd/SIMD.md) commands. This is similar to [<immintrin.h>](https://github.com/gcc-mirror/gcc/blob/master/gcc/config/i386/immintrin.h), which provides wrapper commands and data types for SIMD (e.g. for [Intel](https://software.intel.com/sites/landingpage/IntrinsicsGuide/) ). This method is more effective than  an auto-vectorization tool, because it gives full control to the developer on how to use SIMD in WASM.

**Comment by danchitnis** [comment]
@tlively apologize, didn't realize your [article](https://v8.dev/features/simd) and `<wasm_simd128.h>`. The documentation is a bit fragmented, would be nice to put them in one place!

**Comment by tlively** [comment]
Yes, we will improve the implementation and documentation of `wasm_simd128.h` once the SIMD instruction set is finalized. I'm glad you found it!
